이선민 - 자판기 온보딩 프로젝트

HelloHailie opened the pull request:
### 1차 리뷰 후 todo 리스트

- [x] 린트 및 포매터 강제하도록 수정
- [x] 사용하지 않는 updateBalance 함수 제거
- [x] 자바스크립트에서 디자인 요소 하나하나 만들지 않고 template 사용
- [x] mdn input의 참고해서 valueAsNumber 알아보고 사용
- [x] enter키 사용기능은 form 태그에 submit 함수 사용해서 종속되지 않는 방법으로 웹 표준 방법 사용하기
- [x] 이벤트 리스너 내부 함수 어떤건 전역으로 사용하고 어떤건 아닌데 통일시키기. (추후 테스트를 위해서는 파라미터 사용)

### 추가 todo 리스트

- [x] proxy 사용해서 자동으로 트리거가 될 수 있도록 하기. [참고 링크](https://dev.to/lnahrf/javascript-proxy-magic-how-i-built-a-2kb-state-manager-with-zero-dependencies-and-how-it-got-me-two-different-job-offers-2539)
- [x] 테스트 진행을 위한 모듈화 작업
- [x] 테스트 진행

HelloHailie committed:
Chore: vite 초기 세팅 및 tailwindcss 화면 구성

HelloHailie committed:
Chore: pure css 사용으로 더이상 사용하지 않는 tailwindcss 패키지 제거

HelloHailie committed:
Chore: 사용하지 않는 파일 제거

HelloHailie committed:
Design: 자판기 ui 작업

HelloHailie committed:
Feat: 금액 투입 및 반환 로그 작동 기능 추가

HelloHailie committed:
Feat: 상품 구매 기능 구현 및 관련 로직 추가

- handleProductPurchase 함수 추가: 상품 구매 처리 로직 구현
- 잔액 부족 시 메시지 표시 기능 추가
- 구매 성공 시 잔액 업데이트 및 로그 추가

HelloHailie committed:
Design: 작동 로그에 스크롤 생기도록 추가

HelloHailie committed:
Refactor: 데이터는 products 파일에서 관리하도록 수정

HelloHailie committed:
Fix: 버튼을 누르지 않을 때는 투입된 금액이 표시되도록 수정

Commit:
Fix: 음료 데이터 수정

Commit:
Design: 자판기 금액 투입구와 높이 맞추기

Commit:
Fix: 최신 로그가 아래에 표시되도록 수정

Commit:
Feat: active 상태에 대한 css 추가

Commit:
Fix: 로그 스크롤이 자동으로 스크롤되도록 수정

Commit:
Comments: 길어진 코드에 대해 주석 추가

Commit:
Chore: 자판기 favicon 추가

Commit:
Refactor: 이벤트 버블링을 활용한 이벤트 처리

Commit:
Refactor: 중복되는 현재 잔액 파트 제거 및 css 수정

Commit:
Fix: 자판기 박스의 너비가 화면 너비를 넘지 않도록 수정

Commit:
Docs: README 작성

HelloHailie committed:
Chore: 사용하지 않는 코드 제거

HelloHailie committed:
Refactor:자판기 UI 개선 및 코드 구조 리팩토링

- HTML 구조 개선: 정적 마크업으로 UI 구조 변경
- CSS 스타일 수정: 레이아웃 및 디스플레이 속성 조정
- JavaScript 코드 리팩토링:
  - DOM 요소 참조 방식 개선

HelloHailie committed:
Chore: biome.json 추가 및 리팩터링

taggon commented on src/main.js:
```diff
@@ -0,0 +1,159 @@
+  productGrid: document.querySelector('.product-grid'),
+};
+
+// 유틸리티 함수
```
```suggestion
/**
 * 숫자를 형식에 맞게 변환한다. 세 자리마다 콤마를 추가한다.
 * @param {number} number 숫자
 * @return {string}
 */
```

이왕 주석을 쓸거라면 JSDoc 형태로 해보는 건 어떨까요?

HelloHailie replied:
주요 객체와 함수에 대한 설명을 JSDoc 형태로 작성해보았습니다. 😃

taggon commented on biome.json:
이왕 린터와 포매터를 적용하셨으니 커밋하기 전에 무조건 적용하는 방법도 추가해볼까요?

HelloHailie replied:
husky 사용해서 커밋할때 린터, 포매터 자동 적용하도록 기능 추가하였습니다. 😃

taggon commented on src/main.js:
```diff
@@ -0,0 +1,159 @@
+// 상품 그리드 초기화
+const initializeProductGrid = () => {
+  PRODUCTS.forEach((product) => {
+    const button = document.createElement('button');
```
디자인 요소를 한땀한땀 만드는 것보다는 template 태그를 활용해보는 건 어떨까요? :)

HelloHailie replied:
리더님 말씀해주신 렌더링 시에는 화면에 표시되지 않지만 자바스크립트로 동적 조작이 가능한 template 태그를 적용해보았습니다. 
알려주셔서 감사합니다!!

taggon commented on src/main.js:
```diff
@@ -0,0 +1,159 @@
+
+// 금액 처리 함수
+const handleMoneyInsert = () => {
+  const inputAmount = Number.parseInt(elements.moneyInput.value);
```
`elements.moneyInput.valueAsNumber`는 혹시 확인해보셨을까요?

HelloHailie replied:
input의 type='number'임을 더 잘 활용할 수 있도록 'valueAsNumber' 사용해서 불필요한 형변환 과정을 줄였습니다. 😃

taggon commented on src/main.js:
```diff
@@ -0,0 +1,159 @@
+};
+
+const handleInputValidation = (e) => {
+  e.target.value = e.target.value.replace(/[^0-9]/g, '');
```
`[^0-9]` 대신 `\D`는 어떤가요?

HelloHailie replied:
정규표현식은 관심이 없었는데 쉽게 설명해주셔서 작은 것은 제가 코드 확인하고 검증해보는 식으로 작업하는 습관 들여야겠습니다. 😃

HelloHailie committed:
Chore: Biome 추가 및 husky 설치

HelloHailie committed:
Chore: git commit을 실행하면 Husky가 pre-commit 훅을 실행하도록 수정

HelloHailie committed:
Chore: lint 설정에 따른 추가 커밋

HelloHailie committed:
Fix: 사용하지 않는 코드 제거

HelloHailie committed:
Fix: hover가 아닌 click시 상품 가격 보이도록 수정

HelloHailie committed:
Refactor: 금액 입력 폼 개선 및 이벤트 처리 로직 수정

- 금액 입력 컨테이너를 form 요소로 변경하여 제출 시 기본 동작 방지
- '투입' 버튼을 form의 submit 타입으로 변경

HelloHailie committed:
Refactor: 템플릿 요소를 사용한 상품 그리드 생성 최적화

- HTML에 상품 버튼 템플릿 추가
- JavaScript에서 템플릿을 활용한 상품 그리드 초기화 로직 개선
- DOM 조작 최소화로 성능 향상

HelloHailie committed:
Refactor: 금액 입력 처리 로직 개선

- 입력값 파싱 방식 변경: Number.parseInt() 대신 valueAsNumber 속성 사용
- 유효성 검사 로직 개선: NaN 체크 추가

HelloHailie committed:
Refactor: 금액 입력 유효성 검사 정규식 개선

- 숫자가 아닌 문자 제거 정규식을 /[^0-9]/g에서 /\D/g로 변경

HelloHailie committed:
Refactor: 이벤트 리스너 초기화 로직 개선 및 코드 구조화

- initializeEventListeners 함수에 객체 구조 분해 할당 적용
- initialize 함수에서 필요한 요소만 전달하도록 수정

HelloHailie committed:
Refactor: 금액 투입 시 최대 한도 검사 로직 추가

- 현재 잔액과 투입 금액의 합이 1억원을 초과하는 경우 처리

HelloHailie committed:
Refactor: 코드 문서화 및 가독성 개선

- JSDoc 스타일의 주석 추가로 함수 설명 및 매개변수 타입 명시
- 주요 객체와 함수에 대한 설명 주석 추가

HelloHailie commented on src/main.js:
```diff
@@ -0,0 +1,180 @@
+      updateDisplayAmount(state.currentMoney);
+      break;
+  }
+};
```
@taggon 
이벤트 처리하는 부분을 한 곳에서 처리해서 경우를 나눈 이유가 아래 코드가 중복되서 하나의 함수로 묶어서 진행했는데요, 이정도 규모이면 따로 풀어서 진행하는게 효율적인 방법인걸까요?

`
const target = e.target.closest('.product-button');
  if (!target) return;
  const productId = Number.parseInt(target.dataset.productId);
  const product = PRODUCTS.find((product) => product.id === productId);
`


```
 const handleClick = (e) => {
  const target = e.target.closest('.product-button');
  if (!target) return;
  const productId = Number.parseInt(target.dataset.productId);
  const product = PRODUCTS.find((product) => product.id === productId);
  updateDisplayAmount(product.price);
  handleProductPurchase(product);
};

const handleMouseOut = (e) => {
  const target = e.target.closest('.product-button');
  if (!target) return;
  const productId = Number.parseInt(target.dataset.productId);
  const product = PRODUCTS.find((product) => product.id === productId);
  updateDisplayAmount(state.currentMoney);
};
```

taggon replied:
이런 경우에도 저는 나누는 게 나중에 추적하기가 쉽다고 봐요. 다만 공통된 부분은 다시 별도의 함수로 만들 수 있습니다. 타입도 보고 이해하시라고 편의상 TS로 작성해볼게요.

```ts
function getProductFrom(element: HTMLElement) {
  const button = element?.closest('.product-button');
  const productId = Number.parseInt(button?.dataset.productId);
  const product = productId ? PRODUCTS.find(item => item.id === productId) : null;

  return product;
}

function handleClick(e: MouseEvent) {
  const product = getProductFrom(e.target);
  if (product) {
    updateDisplayAmount(product.price);
    handleProductPurchase(product);
  }
}

function handleMouseOut(e: MouseEvent) {
  const product = getProductFrom(e.target);
  if (product) {
    updateDisplayAmount(state.currentMoney);
  }
}
```

일단 이렇게 수정할 수 있을 것 같기는 한데 이 코드만 보면 `handleMouseOut`에서 product를 가져와야 할 이유가 안 보이네요.

HelloHailie committed:
Chore: vitest 적용 및 초기 설정

taggon commented on src/main.js:
```diff
@@ -0,0 +1,180 @@
+      updateDisplayAmount(product.price);
+      handleProductPurchase(product);
+      break;
+    case 'mouseout':
```
이 경우에는 mouseout 보다는 mouseleave가 더 맞을 겁니다. 그리고 addEventListener의 세 번째 인수를 true로 설정하셔야 할테고요. 차이가 뭘까요?

HelloHailie replied:
오! 궁금해서 찾아봤더니 차이가 명확하게 있네요!
mouseleave 이벤트는 요소의 경계를 완전히 벗어날 때만 발생해서 현재 제 코드는 부모 요소인 productGrid에 이벤트를 걸고 있는데, addEventListener의 세 번째 인수인 useCapture 옵션을 이벤트 캡처링 단계에서 이벤트를 처리한다고 true로 설정해서 하는 방법을 말씀주신 거네요. 이 방법대로 수정해보겠습니다 😃

HelloHailie committed:
Refactor: 그리드 이벤트 핸들러 분리 및 단순화

- handleGridEvent 함수를 handleClick과 handleMouseLeave로 분리
- 제품 조회 로직을 getProductById 유틸 함수로 추출
- mouseout 이벤트를 mouseleave로 변경하여 이벤트 버블링 개선

HelloHailie committed:
Chore: gitignore 정리

- .gitignore 파일을 현대적인 구조로 정리

HelloHailie committed:
Chore: pre-commit 훅 개선 및 lint-staged 설정 수정

- pre-commit 훅을 lint-staged를 사용하도록 변경
- lint-staged 설정에 포맷팅, 린팅, 자동 스테이징 추가

HelloHailie committed:
Refactor: 상태 관리 로직 개선 및 코드 최적화

- Proxy 객체를 사용하여 상태 변경 감지 로직 구현
- 금액 표시 업데이트 로직 중복 제거
- forEach를 for...of로 변경하여 반복문 가독성 개선

HelloHailie committed:
Fix: 상품 템플릿 구조 개선

- product-template을 product-grid 내부로 이동하여 HTML 구조 개선
- 관련 컴포넌트들의 계층 구조를 더 명확하게 정리

HelloHailie committed:
refactor: 코드 구조 개선 및 모듈화

- 기능별로 디렉토리 구조 재구성
  - dom: DOM 요소 조작 및 참조 관련 코드
  - features: 상품과 금액 관련 핵심 기능
  - store: 전역 상태 관리
  - utils: 포맷팅, 로깅 등 유틸리티 함수

- 관련 기능들을 단일 책임 원칙에 따라 모듈화
- 파일 경로 및 import 구문 정리

HelloHailie committed:
Refactor: 불필요한 검증 로직 제거 및 파일명 정리

- moneyService: deductMoney 함수에서 불필요한 금액 검증 제거
- productService: validatePurchase 함수에서 중복 검증 제거
- Product.js → product.js로 파일명 컨벤션 통일

HelloHailie committed:
Test: 테스트 환경 설정 간소화 및 비즈니스 로직 테스트 추가

HelloHailie commented:
@taggon 기존 작업에서는 테스트를 DOM 요소를 모킹하는 식으로 작성해야해서 아래와 같이 main.js 내 함수를 아래를 기준으로 나누어서 비즈니스 로직 부분만 테스트를 작성해보았습니다. 

미진한 부분이 있다면 피드백 부탁드립니다. 🙇


1. controllers/: UI 이벤트 핸들링
2. data/: 정적 데이터
3. dom/: DOM 요소 참조
4. models/: 데이터 모델
5. services/: 비즈니스 로직
6. utils/: 유틸리티 함수들
